### DISCUSSION TITLE
Webpack 5 module federation

### DISCUSSION THREAD

**Discussion by BhaskaranR** [nuxt/nuxt#18079] [General]
How easy would be to do micro frontend with module federation?  According to [https://github.com/sokra/slides/blob/master/content/ModuleFederationWebpack5.md](https://github.com/sokra/slides/blob/master/content/ModuleFederationWebpack5.md). Webpack requires async boundary between the entrypoint and the shared code.

**Comment by phoenix-ru** [comment]
It seems that there is not async boundary on the client, as far as we could progress with implementing MF in Nuxt 3

upd. Also, it seems that the nitro build is interfering with webpack build, causing wrong resolution, such as this (which should be resolved by Module Federation)
`import * as __WEBPACK_EXTERNAL_MODULE_footer_app_7aaa5992__ from 'footer/app';`

**Reply by phoenix-ru** [reply]
I managed to fix the latter issue by attaching to the `webpack:config` hook and rewriting the externals resolution function.
Now the other challenge is that Nuxt builds in `.mjs` mode, causing `require`s from remotes to fail. I am currently trying to befriend our commonjs remotes with modulejs in Nuxt

**Comment by ken-playerlync** [comment]
@phoenix-ru are you working with the nuxt team on this support? I am working on a project, where the next phase is to leverage module federation, however this being a Vue 3 project, I am looking at Nuxt 3, but it looks like I may have to settle for Vue CLI at the moment for my main wrapping component in the micro-frontend architecture we will be implementing.

**Reply by phoenix-ru** [reply]
No, I am not related to Nuxt team.
At our company we have managed to implement module federation in Vue 3 CLI + SSR.
Since most of our projects rely on Nuxt 2, we wanted to utilize Nuxt 3 for a smooth transition, integrating federated remote components along the way. Unfortunately, we are currently struggling with the lack of documentation in Nuxt 3.

**Reply by danielroe** [reply]
Feel free to ask ☺️ We're still writing things down and asking good questions does help us as we do so...

**Reply by phoenix-ru** [reply]
@danielroe In case you know, how do we force Nuxt and Nitro to build CommonJS modules?
Interfering with `output.libraryTarget`, etc. seems to break Nitro build, as Vue.js and Vue server renderer are not being included in the rollup build, no matter whether you specify Vue as shared or not.

Unfortunately, I have no idea how to build .mjs remotes for consumption in Node.js without NodeAsyncHttpRuntime by @telenko/node-mf.

**Reply by ken-playerlync** [reply]
@phoenix-ru yeah, I am currently spiking and per documentation from Vue 3, it is recommended to use Vue CLI over Vite, if something like federated modules is needed. So then I started looking at Nuxt 3, since I have used Nuxt 2, but alas finding this discussion, looks like I cannot use Nuxt 3 either. Thanks for the response.

**Reply by danielroe** [reply]
Out of curiosity, why do you need to generate Common JS? The module dependencies in this repo are ESM and it's important that the Nitro build is also ESM.

**Reply by phoenix-ru** [reply]
The reason for CommonJS is the way Module Federation works in Node.js. Currently it is supported via NodeModuleFederatio plugin written by @telenko. The code generated by the plugin needs access to __dirname and __filename, on top of using require('fs') and require('vm').

**Reply by danielroe** [reply]
We can likely emulate those using https://github.com/unjs/mlly:

**shim**:
```js
import { createCommonJS } from 'mlly'

const { __dirname, __filename, require } = createCommonJS(import.meta.url)
```

**Comment by phoenix-ru** [comment]
I managed to get Module Federation working on Nuxt 3 SSR 🌟

The only step needed for a full support is rewriting the entrypoint for the client to use async import. I will create an issue for that

**Reply by danielroe** [reply]
Amazing! That's great to hear. Would love to see your code and make this as seamless as possible with Nuxt 3 🎉

**Reply by phoenix-ru** [reply]
Thanks :)

There is some underlying work to get this setup working. It turns out that the remote component I tested on uses my custom `vue-style-loader` and `vue-loader` adapted for microfrontends in order to inline styles the same way this was done in Vue 2.
I am not sure about sustainability of this approach and would like to know how was the critical CSS collection issue solved in Nuxt 3? Unfortunately, I haven't succeeded with manifests approach, as the federated application has multiple entries and we would need to attach CSS-handling code everywhere.

**Comment by phoenix-ru** [comment]
Issue opened with reproduction
https://github.com/nuxt/framework/issues/1444

**Comment by oliver-heeger** [comment]
@phoenix-ru any idea how to fix `Uncaught Error: Shared module is not available for eager consumption: webpack/sharing/consume/default/vue/vue`. This also happens in your Example.

**Reply by phoenix-ru** [reply]
This is exactly the topic of an issue (link above). Module Federation on client will start working when the issue is resolved

**Comment by phoenix-ru** [comment]
Solved an issue locally. Will provide a PR soon

**Reply by phoenix-ru** [reply]
In case anyone is interested, PR is here https://github.com/nuxt/framework/pull/2223
I am still waiting for maintainers to merge it.

**Reply by erisonalexandre** [reply]
Thanks, I need this functionality, but I need to integrate with projects in Vue 2, but it's giving that I can't find the component's _init.

**Reply by phoenix-ru** [reply]
You can't mix Vue 2 and Vue 3, no matter if it's Module Federation or Nuxt.
But what you can do is either recompile your projects to Vue 3 or use a solution like vue-demi (which might be tedious)
